homeassistant/components/media_player/cast.py

```diff
@@ -33,90 +38,164 @@
+            _LOGGER.debug("Can't set up cast with mDNS info %s. "
+                          "Assuming it's not a Chromecast", mdns)
             return
+        hass.data[KNOWN_CHROMECASTS_KEY][key] = chromecast
+        dispatcher_send(hass, SIGNAL_CAST_DISCOVERED, chromecast)
 
-        hosts = [host]
+    _LOGGER.debug("Starting internal pychromecast discovery.")
+    listener, browser = pychromecast.start_discovery(internal_callback)
+    hass.data[INTERNAL_DISCOVERY_RUNNING_KEY] = True
 
-    elif CONF_HOST in config:
-        host = (config.get(CONF_HOST), DEFAULT_PORT)
+    def stop_discovery():
+        """Stops discovery of new chromecasts."""
+        pychromecast.stop_discovery(browser)
+        hass.data[INTERNAL_DISCOVERY_RUNNING_KEY] = False
 
-        if host in known_hosts:
-            return
+    hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, stop_discovery)
+
+
+@callback
+def _async_create_cast_device(hass, chromecast):
+    """Create a CastDevice Entity from the chromecast object.
 
-        hosts = [host]
+    Returns None if the cast device has already been added. Additionally,
+    automatically updates existing chromecast entities.
+    """
+
+    if chromecast.uuid is None:
+        # Found a cast without UUID, we don't store it because we won't be able
+        # to update it anyway.
+        return CastDevice(chromecast)
+
+    # Found a cast with UUID
+    known_casts = hass.data[ADDED_CAST_DEVICES_KEY]
+    old_cast_device = known_casts.get(chromecast.uuid)
+    if old_cast_device is None:
+        # -> New cast device
+        cast_device = known_casts[chromecast.uuid] = CastDevice(chromecast)
+        return cast_device
+
+    # -> Cast device changed host
+
+    # Remove old pychromecast.Chromecast from global list, because it isn't
+    # valid anymore
+    key = (old_cast_device.cast.host,
+           old_cast_device.cast.port,
+           old_cast_device.cast.uuid)
+    hass.data[KNOWN_CHROMECASTS_KEY].pop(key)
+    old_cast_device.async_set_chromecast(chromecast)
+    return None
+
+
+@asyncio.coroutine
+def async_setup_platform(hass: HomeAssistantType, config: ConfigType,
+                         async_add_devices, discovery_info=None):
+    """Set up the cast platform."""
+    import pychromecast
+
+    # Import CEC IGNORE attributes
+    pychromecast.IGNORE_CEC += config.get(CONF_IGNORE_CEC, [])
+    hass.data.setdefault(ADDED_CAST_DEVICES_KEY, {})
+    hass.data.setdefault(KNOWN_CHROMECASTS_KEY, {})
 
+    # None -> use discovery; (host, port) -> manually specify chromecast.
+    want_host = None
+    if discovery_info:
+        want_host = (discovery_info.get('host'), discovery_info.get('port'))
+    elif CONF_HOST in config:
+        want_host = (config.get(CONF_HOST), DEFAULT_PORT)
+
+    enable_discovery = False
+    if want_host is None:
+        # We were explicitly told to enable pychromecast discovery.
+        enable_discovery = True
+    elif want_host[1] != DEFAULT_PORT:
```

Comment:
Doesn't this fail if the discovery info in line 143 sets the port to `DEFAULT_PORT` by chance?

Reply:
That's the whole point of this line! We only need to run discovery if we find a group (and groups always have non-standard ports).

So if `discovery:` finds a chromecast with `DEFAULT_PORT`, we don't need to run our own discovery.

Reply:
Okay, if groups are guaranteed to live on different ports, that should work. 👍